### PR TITLE
[PrunedLiveness] Fix boundary check for dead-ends.

### PR DESCRIPTION
**Pull request by nate-chandler**
Consider dead, dead-end blocks within the availability boundary to be within the boundary.

rdar://126965232

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please apple silicon benchmark

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
Source compat failures match failures in baseline:
```
Failures:
  FAIL: penny-bot, 5.9, 62006b, Swift Package
  FAIL: swift-sdk-generator, 5.9, e40e1f, Swift Package
```

File: lib/SIL/Utils/PrunedLiveness.cpp [hunk 1]
```diff
@@ -489,8 +489,50 @@ bool PrunedLiveRange<LivenessWithDefs>::isWithinBoundary(
...
+  if (liveBlocks.getBlockLiveness(parent) != PrunedLiveBlocks::Dead) {
+    return false;
+  }
+  // Check whether the value is available in `parent` (i.e. not consumed on any
+  // path to it):
+  //
+  // Search backward until LiveOut or LiveWithin blocks are reached.
+  // (1) If ALL the reached blocks are LiveOut, then `parent` IS within the
+  //     boundary.
+  // (2) If ANY reached block is LiveWithin, then `parent` is NOT within the
```
Comment (atrick) [inline lib/SIL/Utils/PrunedLiveness.cpp:512]: Do you think it's obvious enough that `LiveWithin` implies that this path reached a consume, and therefore makes the value unavailable?

**Review by atrick — APPROVED** [review]
LGTM

File: lib/SIL/Utils/PrunedLiveness.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SIL/Utils/PrunedLiveness.cpp]: Expanded the comment to
```
(2) If ANY reached block is LiveWithin, the value was consumed in that
    reached block, preventing the value from being available at `parent`,
    so `parent` is NOT within the availability boundary.
```
and added backreferences to (1) and (2) from the switch below.

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

**Pull request merged**